### PR TITLE
Feature/add segformer family

### PR DESCRIPTION
**Pull request by imagra93**
# Add SegFormer semantic segmentation family (LibreSegformer, b0-b5)

## Summary

Adds `LibreSegformer`, a native semantic-segmentation family covering all six
standard sizes (b0 through b5): a Mix Transformer (MiT) encoder with an all-MLP
decode head. The family is registered, trainable, validatable, and ships
pretrained ADE20K weights for every size.

Reference: SegFormer, Xie et al., NeurIPS 2021.

## What this adds

- `libreyolo/models/segformer/`: model wrapper, `nn.py` (MiT encoder + decode
  head), `trainer.py`, and a family `NOTICE`.
- Registration in `libreyolo/__init__.py` and `libreyolo/models/__init__.py`.
- `SegformerConfig` training defaults (AdamW, backbone base LR with a decode
  head LR multiplier, linear decay, scale-jitter, ImageNet-normalized inputs).
- `weights/convert_segformer_weights.py` and hosted weights for all six sizes.
- `tests/unit/test_segformer.py` (39 tests).
- Contract docs: `docs/nomenclature.md`, `NOTICE`, `THIRD_PARTY_NOTICES.txt`,
  `weights/LICENSE_NOTICE.txt`, and the HF upload skill's filename whitelist.

## Code provenance

The architecture (`libreyolo/models/segformer/nn.py`) is a native PyTorch
re-implementation derived from HuggingFace Transformers' **Apache-2.0**
`models/segformer/{configuration_segformer.py, modeling_segformer.py}`
(Copyright 2021 NVIDIA and The HuggingFace Inc. team). There is no runtime
dependency on the `transformers` package.

NVIDIA's original `NVlabs/SegFormer` repository is under the non-commercial
NVIDIA Source Code License. Its **code was not read, copied, or derived from**;
it is cited for paper attribution only. Where the Apache-2.0 reference and the
original differ (e.g. LayerNorm eps at the block/stage norms), LibreYOLO follows
the **Apache-2.0 reference**, deliberately, to keep provenance clean. That choice
is verified to cost nothing: ADE20K mIoU is identical either way.

The `-sem` **weights** are a separate matter and are NOT Apache-2.0 (see below).

## Weights (NON-COMMERCIAL)

The pretrained ADE20K checkpoints are converted from NVIDIA's release
(`nvidia/segformer-b0..b5-finetuned-ade-*`). The NVIDIA Source Code License
**permits redistribution** of the work and of derivative works, provided a
complete copy of the license travels with them and attribution notices are
retained (Sections 2.1, 3.1), but it **limits use to non-commercial research or
evaluation** (Section 3.3), virally through derivative works (Section 3.2).

So LibreYOLO hosts them, clearly labelled, exactly as it does the VisDrone
research-preview weights:

- Each HF repo ships the verbatim `LICENSE`, a `NOTICE`, and
  `license: other` + `license_name` + `license_link` metadata.
- `LibreSegformer.get_download_notice` prints the restriction **before** every
  auto-download.
- The architecture, LibreYOLO's code, and any model you train from scratch are
  unaffected and remain permissively licensed.

Conversion is a state-dict key remap only (`segformer.` -> `encoder.`); learned
parameters are NVIDIA's, unchanged.

## Verification

- **Bit-exact parity** with the Apache-2.0 reference: max abs logit difference
  **0.0** in float64 for b0-b4, through the shipped converter, strict-loaded.
- Parameter counts match the published table exactly for all six sizes
  (3.75 / 13.72 / 27.46 / 47.34 / 64.11 / 84.71 M).
- **Real ADE20K val (2000 images)**: b0 **36.5** mIoU, b1 **40.7** (paper
  single-scale: 37.4 / 42.2; the gap is LibreYOLO's fixed-canvas letterbox eval
  protocol, not a defect).
- Cold auto-download from HF verified end to end for a fresh install.
- From-scratch training converges (semantic pipeline, `resize_crop` sampling,
  validator, checkpoint round-trip).
- 39 unit tests, plus the semantic/classify/EoMT/PIDNet suites, pass unchanged.

## Maintainer fixes on top of the original branch

The architecture port was faithful except for three constants, all now pinned by
tests:

- **LayerNorm eps** was 1e-6 at all five sites; the reference uses torch's
  default 1e-5 (`layer_norm_eps` in the upstream configs is read by nothing).
  Broke bit-exactness; no accuracy impact.
- **Mix-FFN dropout** was 0.1 (the decode head's value); the reference uses
  `hidden_dropout_prob` = 0.0. Training-time only.
- **No weight init**, so torch's kaiming-uniform default applied (~5x too wide
  at the narrow stages, non-zero biases). Added the reference init.

Plus real bugs:

- `_load_weights` never rebuilt the net for the checkpoint's **size**, so any
  non-default size (`LibreSegformerb1-sem.pt` into the b0 default) died on shape
  mismatches. Now follows EoMT's `_rebuild_for_new_size` pattern.
- **b5 is fine-tuned at 640**, not 512; `INPUT_SIZES` hardcoded 512 for all six.
- A tuple `imgsz` hit the divisor check before it could be widened (Greptile).

`tools/pretrain_mit/` and `scripts/train_segformer_all_sizes.py` were removed:
pretraining MiT from scratch on ImageNet existed only to work around weights we
are in fact allowed to ship.

## Known limitations

- Export is not implemented (`export()` raises), matching PIDNet and EoMT; the
  semantic task has no export support anywhere yet.
- Test-time augmentation is not supported for semantic models.

### CODE REVIEW

**Comment by EHxuban11** [comment]
@greptileai review

Pushed maintainer fixes on top: reference-fidelity constants (LayerNorm eps, Mix-FFN dropout, weight init), a size-rebuild fix in `_load_weights`, b5 input size 640, the rectangular-imgsz guard you flagged, and the ADE20K weights are now converted and hosted. `tools/pretrain_mit/` and `scripts/train_segformer_all_sizes.py` are removed.

File: libreyolo/models/segformer/model.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+    def _preprocess(
+        self,
+        image: ImageInput,
+        color_format: str = "auto",
+        input_size: Optional[int] = None,
+    ) -> Tuple[torch.Tensor, Image.Image, Tuple[int, int], float]:
+        effective_res = input_size if input_size is not None else self._get_input_size()
+        divisor = self.semantic_imgsz_divisor
+        if any(side % divisor for side in _input_size_hw(effective_res)):
```
[the remark this answers is not in the document]
Comment (EHxuban11) [reply libreyolo/models/segformer/model.py:269]: Checked this empirically and it does not reproduce: it is unreachable for this family.

`_predict_tiled` short-circuits before the comparison you flagged:

```python
# libreyolo/models/base/inference.py:1102-1105
# Tiling is a detection-time technique; for whole-image classification
# and dense semantic maps it is meaningless, so fall back to a single
# forward pass.
if getattr(self.model, "task", None) in ("classify", "semantic"):
    return self._predict_single(...)
```

LibreSegformer is `task="semantic"`, so it returns from that guard at line 1105 and never reaches the `orig_width <= input_size` check at line 1148. Verified against the real ADE20K weights, with the tiled path instrumented to confirm it is entered:

```
>>> _predict_tiled ENTERED, imgsz= (512, 768)
tuple+tiling -> semantic_mask: True
```

`predict(tiling=True)`, `predict(imgsz=(512,768))`, and both together all return a correct mask, no TypeError.

The underlying int-vs-tuple comparison at line 1148 is real but belongs to the shared detection tiling path, and `predict()` types `imgsz` as `Optional[int]`, so a tuple is not a supported public input there either. Out of scope for this PR; not introducing a fix for an unreachable path.

Your earlier rectangular-imgsz finding was correct and is fixed (`_input_size_hw` now runs before the divisor check).
